prng.c

```diff
@@ -3,27 +3,97 @@
+
+static inline prng_timer_t timer_start(void) {
+  return _rdtsc();
+}
+static inline int64_t timer_elapsed(prng_timer_t start) {
+  int64_t end = _rdtsc();
+  // While latency isn't as much of a problem with _rdtsc as it is with
+  // clock_gettime, measure it anyways to be safe.
+  static int64_t latency = -1;
+  if (latency == -1) {
+    const int LATENCY_ROUNDS = 256;
+    latency = 0;
+    // sample 256 times and get the average.
+    for (int i = 0; i < LATENCY_ROUNDS; i++) {
+      int64_t a = timer_start();
+      int64_t b = timer_start();
+      latency += b - a;
+    }
+    latency /= LATENCY_ROUNDS;
+  }
+  return end - start - latency;
+}
+#else
+// fall back to nanoseconds, e.g. on aarch64 where __builtin_readcyclecounter
+// needs special privledges not usually granted by the Linux kernel.
+// Note that this depends on the CPU frequency.
+// This assumes that clock_gettime is available.
+#include <time.h>
+
+typedef struct timespec prng_timer_t;
+static const char *unit = "ns/byte";
+
+static inline prng_timer_t timer_start(void) {
+   struct timespec ret;
+   clock_gettime(CLOCK_REALTIME, &ret);
+   return ret;
+}
+// Converts a struct timespec to nanoseconds
+static inline int64_t timespec_to_ns(struct timespec spec) {
+  return (int64_t)spec.tv_nsec + ((int64_t)spec.tv_sec * (int64_t)1e9);
+}
+static inline int64_t timer_elapsed(prng_timer_t start) {
+  struct timespec end = timer_start();
+  int64_t start_ns = timespec_to_ns(start);
+  int64_t end_ns = timespec_to_ns(end);
+  // Measure the latency of clock_gettime.
+  // Since it is sensitive to clock speed and such, we measure each time.
+  int64_t latency = 0;
+  const int LATENCY_ROUNDS = 64;
+  // sample 64 times and get the average.
+  for (int i = 0; i < LATENCY_ROUNDS; i++) {
+    struct timespec a = timer_start();
+    struct timespec b = timer_start();
+    latency += timespec_to_ns(b) - timespec_to_ns(a);
+  }
+  latency /= LATENCY_ROUNDS;
+  return end_ns - start_ns - latency;
+}
+#endif
 #define BUFSIZE (1<<17)
 #define SEEDTYPE uint64_t
-#include "./prng.h"
-typedef struct args { int64_t bytes; SEEDTYPE seed[4]; int rval; } args_t;
+#ifndef HEADER
+#  define HEADER "./prng.h"
+#endif
+#include HEADER
+
+typedef struct args { int64_t bytes; SEEDTYPE seed[4]; int rval; int quiet; } args_t;
 args_t parseArgs(int argc, char **argv);
 
 int main(int argc, char **argv) {
   args_t a = parseArgs(argc, argv);
   if (a.rval < 0) { return a.rval; }
   prng_state s = prng_init(a.seed);
   uint8_t buf[BUFSIZE] __attribute__ ((aligned (64)));
-  int64_t cycles = 0, start;
+  int64_t cycles = 0;
+  prng_timer_t start;
   for (int64_t bytes = a.bytes; bytes >= 0; bytes -= sizeof(buf)) {
     int wbytes = bytes < sizeof(buf)? bytes: sizeof(buf);
-    start = _rdtsc();
+    start = timer_start();
     prng_gen(&s, buf, sizeof(buf));
-    cycles += _rdtsc() - start;
-    ssize_t w = write(STDOUT_FILENO, buf, wbytes);
+    cycles += timer_elapsed(start);
```

Comment:
The cpb have changed a bit, but I'm not 100% sure why.

Reply:
Is it faster or slower?

Reply:
Take look to [this comment](https://github.com/rurban/smhasher/issues/125#issuecomment-625776324), it may be useful for measuring CPU clock cycles.

Reply:
With the current code, it is less precise. We used to have two digits of precision (so, very reproducible results), and now it is ±0.05 cpb. Also, SHISHUA and xoshiro256+x8 is slower, ChaCha8 and a few others are faster.

Reply:
@erthink when measuring bytes/cycle in code involving AVX, both rdtscp and my approach will fail - rdtscp in intel cpus measures ticks of basic frequency (f.e. it's ~2 GHz for i9-7820 and i7-8665), while my approach employing only scalar ops indeed measures frequency in AVX-less code, while AVX frequency may be lower

Reply:
Proper use of rdtscp described in https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/ia-32-ia-64-benchmark-code-execution-paper.pdf

(in short: rdtscp + code + rdtscp + save eax/adx/... + cpuid), cpuid ensures that later commands will not started executing before cpuid so rdtscp will count only time necessary to execute code between two rdtscp commands)

Reply:
Hmm, from what I am reading, we don't need to do cpuid, only mfence and lfence, or even better, use rdtscp:
We should be able to do this
```c
static inline int64_t timer_start(void)
{
#ifdef __GNUC__
    size_t eax, edx; // avoid manual zero extending on 64-bit
    __asm__ __volatile__(
        "mfence\n"
        "lfence\n"
        "rdtsc"
      : "=a" (eax), "=d" (edx) :: "memory");
    return eax | ((int64_t)edx << 32);
#else
    _mm_mfence();
    _mm_lfence();
    return _rdtsc();
#endif
}

static inline int64_t timer_end(int64_t start)
{
#if defined (__SSE4_2__) || defined(__AVX__)
    // rdtscp is only available on Nehalem and later.
#  ifdef __GNUC__
    size_t eax, edx;
    __asm__ __volatile__(
        "rdtscp\n"
        "lfence"
        : "=a" (eax), "=d" (edx) :: "%ecx", "memory");
    int64_t end = eax | ((int64_t)edx << 32);
#  else
    unsigned dummy;
    int64_t end = _rdtscp(&dummy);
    _mm_lfence();
#  endif
    return end - start;
#else 
    return timer_start() - start;
#endif
}

```

Reply:
mfence/lfence is MEMORY OPERATION fences while rdtscp+cpuid makes fences around ALL COMMANDS executed by the cpu. this ensures that you measure only operations executed between rdtscp commands.